File: 14-02-view-group/java-14-02-view-group/app/src/main/java/com/example/viewgroup/ToastEditText.java
package com.example.viewgroup;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.Toast;


public class ToastEditText extends androidx.appcompat.widget.AppCompatEditText {

    private Toast toast;

    public ToastEditText(Context context) {
        super(context);
    }

    public ToastEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public ToastEditText(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void onTextChanged(CharSequence text, int start, int lengthBefore, int lengthAfter) {
        super.onTextChanged(text, start, lengthBefore, lengthAfter);
        if (toast != null) {
            toast.cancel();
        }
        toast = Toast.makeText(getContext(), "length is " + getText().length(), Toast.LENGTH_SHORT);
        toast.show();
    }
}

File: 14-01-subclass-view/kotlin-14-01-subclass-view/app/src/main/java/com/example/subclassview/ToastEditText.kt
package com.example.subclassview

import android.content.Context
import android.util.AttributeSet
import android.widget.Toast


class ToastEditText : androidx.appcompat.widget.AppCompatEditText {

    private var toast: Toast? = null

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    override fun onTextChanged(text: CharSequence, start: Int, lengthBefore: Int, lengthAfter: Int) {
        super.onTextChanged(text, start, lengthBefore, lengthAfter)
        if (toast != null) {
            toast?.cancel()
        }
        toast = Toast.makeText(context, "length is " + getText()?.length, Toast.LENGTH_SHORT)
        toast?.show()
    }
}
